### ISSUE TITLE
FitHiC

### ISSUE THREAD

**Issue by trytrytry1369** [Bioconductor/Contributions#141]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/trytrytry1369/FitHiC

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by mtmorgan** [comment]
Thank you for your contribution

- Please expose the functionality (input and return values) using
  Bioconductor classes (e.g., from the GenomicRanges and / or
  InteractionSet pacakges) , instead of or in addition to the
  file-based functions.
- Justify use of idiosyncratic file representations rather than
  'standard' (gff, bed, ...) formats. Using existing tools (e.g.,
  `rtracklayer::import()`) for file parsing, where appropriate. Even
  internally, carefully consider the need for `data.table`, with it's
  non-standard pass-by-reference semantic.
- reconsider use of C++ for code that is easily represented as
  vectorized calcuations in R. For instance
  [calculate_probabilities_helper][] seems to be simple vector
  arithmetic. Writing in R will be nearly as performant, and the
  algorithm will be transparently accessible to R programmers /
  users. Likewise `benjamini_hochberg_correction_helper`.
  
DESCRIPTION

- Provide only author names / (optional for all but one) email
  addresses in the Author field, consider using the Authors@R syntax
  to designate different 'roles' including the orginal
  author. Describe author roles in detail in the Description field or
  a man page for the package, FitHiC-package.Rd
- Use the Desciprtion: field to highlight the unique contributions of
  the package, much like an abstract in a scientific paper.

vignettes

- consider wrapping lines to a standard character width (80).
- E.g., line 65. Format code chunks to no more than 80 column
  width. Avoid nested function calls that make it difficult to
  associate arguments with their value.
- consider naming code chunks (`r my-chunk, ...`) so that errors are
  easily referenced.

R

- use `^` rather than (undocumented) `**` for exponentiation
- use `message()` rather than `print()` to report progress to users;
  do so optionally via a `verbose=FALSE` argument.
- e.g., FitHiC.R:153 always return a useful (computable) value.
- e.g., FitHiC.R:125 avoid `seq()` in favor of `seq_len()` or
  `seq_along()` or other approaches to guard agains problems of
  reverse sequences, e.g., `seq(2, 1)`
- Rather than returning a plain `list()` of return values, create a
  light-weigth S4 class with defined fields and accessor functions to
  validate 
- Expose visualization capabilities as stand-alone functions so that
  users may call these on computed results without re-running the
  entire work flow. Allow these to be viewed interactively rather than
  creating files.
- use `file.path()` rather than `paste()` to create file paths. 

man

- consider adding a file describing the package, including detailing
  the contributions of each author.
- avoid nested function calls (e.g., `FitHiC(system.file(...))`)to
  clarify the relationship between arguments and their values.

inst/NEWS

- provide content (preferred) or remove this file.

[calculate_probabilities_helper]: https://github.com/trytrytry1369/FitHiC/blob/master/src/helper.cpp#L7

**Comment by lawremi** [comment]
Do you really need data.table here? There could be a good reason (i.e., performance), but all usage looks like simple manipulations that would be fast enough with data.frame. Just curious.

**Comment by trytrytry1369** [comment]
@lawremi There is a big performance difference between data.table and data.frame. I tried data.frame at the beginning but the run time was too long.

**Comment by lawremi** [comment]
It would be interesting to know in which situation that performance difference arises.

**Comment by trytrytry1369** [comment]
@lawremi I think data.table performs better in the situation where data sets do lots of inner join, outer join, cross join, etc.

**Comment by trytrytry1369** [comment]
Hello @mtmorgan 
Thanks for your review. In the latest version(0.99.1), most of issues in DESCRIPTION, vignettes, R, man and inst/NEWS have been fixed.
For the issues related to Bioconductor classes, file formats and interactive visualization, I have to talk with my mentor Ferhat Ay. Since this package is translated from his Python original https://noble.gs.washington.edu/proj/fit-hi-c/, the work flow, input and output are consistent with the Python script.
For the issue related to the use of C++, I tried to write calculate_probabilities_helper and benjamini_hochberg_correction_helper in R but met efficiency problem. That is why I choose C++ for them.

**Comment by mtmorgan** [comment]
For BH, something like (completely untested!)

```
benjamini_hochberg_correction_helper <-
    function(p_values, num_total_tests)
{
    stopifnot(
        is.numeric(p_values),
        is.integer(num_total_tests), length(num_total_tests) == 1L)

    order <- order(p_values)
    sorted_pvals <- p_values[order]

    bh_value <- ifelse(
        p_value == 1L, 1L,
        pmin(p_value * num_total_tests / (seq_along(p_value) + 1)), 1)

    cummax(bh_value)[order]
}
```

where `ifelse()`, `pmin()`, `cummax()` all help with vectorization. But maybe you're really looking for `?p.adjust`? For the other function I couldn't really see what part of the implementation would require iteration rather than vectorization? If you could provide a simple example of where the stumbling block is I'd be happy to provide a solution...

**Comment by trytrytry1369** [comment]
Hi @mtmorgan 
Thanks for your hint about re-writing BH method in R! I know the method `p.adjust(p, method = p.adjust.methods, n = length(p))` but it cannot handle the condition where `n < length(p)`. That is why I write the BH method in C++. But it seems your solution for BH method in R is elegant so I modify my code and submit it as Version 0.99.2.
For calculate_probabilities_helper, the difficulty for me to write it in R is how to separate the data into several bins. For example, the data set is `d1, d2, d3, d4, d5, ... , dn`, I need to separate it to `(d1, d2, d3)`, `(d4, ..., dn)` when `(d1, d2, d3)` satisfies the condition:
1. at least one in `(d1, d2, d3)` outlier should be 0.
2. the sum of hit count of `(d1, d2, d3)` just exceeds the specified threshold.
I appreciate if you can provide me some hint.

**Comment by trytrytry1369** [comment]
Hi @mtmorgan 
For the issue related to file representation, I talked with Ferhat Ay, author of Python original, and he said using current format is enough. His reasons are below as quoted
> As I mentioned, we do not want to do that because a) it is waste of space, b) many people used fithic as is now so we don't want to change the format

BTW, the build of FitHiC is still in `ERROR` because of not subscribing bioc-devel mailing list. However, I have registered for several times but it seems that nobody reply me. Could you help me for this problem? Thanks!

**Comment by mtmorgan** [comment]
Are you subscribing using the email address in the DESCRIPTION file? Also, I believe you'll be sent a confirmation email, maybe it's being marked as spam?

**Comment by mtmorgan** [comment]
Thank you for your revisions. I did not immediately see a way to effectively implement your other C++ function in R.

I will mark your package as accepted. It will be added to the nightly builds and svn repository, with additional information sent in the next several days.

AFTER the package has been added to SVN, I suggest you remove the `benjamini_hochberg_correction_helper()` C++ code, to keep the code base clean and relevant.
